tango/common/from_params.py

```diff
@@ -141,7 +141,21 @@ def infer_method_params(
             # `from __future__ import annotation`, the annotation will be a str,
             # so we need to resolve it using `get_type_hints` from the typing module.
             # See https://www.python.org/dev/peps/pep-0563/ for more info.
-            parameters[param_name] = param.replace(annotation=get_type_hints(method)[param_name])
+            try:
+                parameters[param_name] = param.replace(
+                    annotation=get_type_hints(method)[param_name]
+                )
+            except TypeError as e:
+                if "'type' object is not subscriptable" in str(e):
+                    new_e = TypeError(
+                        f"Could not determine the type annotations for `{method.__qualname__}`. Did you use the "
+                        "lowercase `dict`, `list`, `set`, or `tuple` in the type annotations instead of the "
+                        "uppercase `Dict`, `List`, `Set`, `Tuple`?"
+                    )
```

Comment:
How about:
```suggestion
                if "'type' object is not subscriptable" in str(e):
                    # This can happen when someone uses a type hint like `dict[str, str]`
                    # instead of `Dict[str, str]`.
                    err_msg = (
                        f"Failed to parse the type annotation `{param.annotation}` "
                        f"from `{cls.__qualname__}.{method.__name__}()`."
                    )

                    if "[" in param.annotation:
                        # Check if there is an equivalent generic in the `typing` module.
                        import typing

                        type_, *_ = param.annotation.split("[", 1)
                        for possible_typing_equivalent in {type_, type_.title()}:
                            if hasattr(typing, possible_typing_equivalent):
                                err_msg += (
                                    f" Try using `{possible_typing_equivalent}` "
                                    "from the `typing` module instead."
                                )
                                break

                    new_e = TypeError(err_msg)
                    new_e.__cause__ = e
```

This would result in an error message like:

```
TypeError: Failed to parse the type annotation `dict[str, str]` from `Foo.__init__()`. Try using `Dict` from the `typing` module instead.
```

Reply:
I like it.

Reply:
What's the difference between `{cls.__qualname__}.{method.__name__}` and `{method.__qualname__}`?

Reply:
`method.__qualname__` will look like unrelated garbage for dataclasses, unfortunately. E.g. `__create_fn__.<locals>.__init__`.